crates/kornia-pnp/src/ransac.rs

```diff
@@ -0,0 +1,484 @@
+//! RANSAC-based robust wrapper for PnP solvers.
+
+use crate::ops::{intrinsics_as_vectors, pose_to_rt, project_sq_error};
+use crate::pnp::{PnPError, PnPResult};
+use crate::{solve_pnp, PnPMethod};
+use glam::{Mat3, Vec3};
+use log::{debug, warn};
+use rand::seq::SliceRandom;
+use rand::{rngs::StdRng, SeedableRng};
+
+const MIN_CORRESPONDENCES: usize = 4; // Minimum 2D-3D pairs required by PnP
+const EPNP_MIN_SAMPLE_SIZE: usize = 5; // Minimal sample size for EPnP (unless only 4 points available)
+
+const DEFAULT_MAX_ITERATIONS: usize = 100;
+const DEFAULT_REPROJ_THRESHOLD_PX: f32 = 8.0;
+const DEFAULT_CONFIDENCE: f32 = 0.99;
+
+const EPS_PROB_MIN: f32 = 1e-6; // Guard for tiny probabilities
+const EPS_LOG_GUARD: f32 = 1e-12; // Guard to avoid log(0) and log(1)
+const HIGH_INLIER_RATIO_STOP: f32 = 0.95; // Early stop when inlier ratio is very high
+
+/// Parameters for RANSAC over PnP.
+#[derive(Debug, Clone)]
+pub struct RansacParams {
+    /// Maximum number of RANSAC iterations.
+    pub max_iterations: usize,
+    /// Pixel error threshold to classify an observation as an inlier.
+    pub reproj_threshold_px: f32,
+    /// Desired probability that at least one sample set is outlier-free.
+    pub confidence: f32,
+    /// Optional fixed seed for reproducible sampling.
+    pub random_seed: Option<u64>,
+    /// Whether to refit on all inliers using the base solver.
+    pub refine: bool,
+}
+
+impl Default for RansacParams {
+    fn default() -> Self {
+        Self {
+            max_iterations: DEFAULT_MAX_ITERATIONS,
+            reproj_threshold_px: DEFAULT_REPROJ_THRESHOLD_PX,
+            confidence: DEFAULT_CONFIDENCE,
+            random_seed: None,
+            refine: true,
+        }
+    }
+}
+
+/// RANSAC result for PnP.
+#[derive(Debug, Clone)]
+pub struct PnPRansacResult {
+    /// Best pose found by RANSAC.
+    pub pose: PnPResult,
+    /// Indices of inlier correspondences.
+    pub inliers: Vec<usize>,
+}
+
+/// Solve PnP robustly using a legacy RANSAC loop around a base PnP method (e.g., EPnP).
+///
+/// - Minimal sample size is 5 for EPnP (4 when only 4 points available).
+/// - Scoring uses Euclidean pixel reprojection error.
+/// - Iterations adapt from current inlier ratio and desired confidence.
+pub fn solve_pnp_ransac(
+    world: &[[f32; 3]],
+    image: &[[f32; 2]],
+    k: &[[f32; 3]; 3],
+    base: PnPMethod,
+    params: &RansacParams,
+) -> Result<PnPRansacResult, PnPError> {
```

Comment:
shouldn't we return a Restult<Optional<...>> ? to be very clear when ransanc or just pnp failed or not converged

Reply:
I've added a new `PnPRansacError` enum, can you check?